pkg/network/vlan/vlan.go

```diff
@@ -2,190 +2,130 @@ package vlan
-	br := iface.NewBridge(BridgeName)
+func NewVlan(name string, vids []uint16) *Vlan {
+	br := iface.NewBridge(iface.GenerateName(name, iface.BridgeSuffix))
+
 	return &Vlan{
-		bridge:      br,
-		status:      &network.Status{},
-		eventSender: eventSender,
-		mgmtNetwork: mgmtNetwork,
-		vids:        vids,
+		name:   name,
+		bridge: br,
+		vids:   vids,
 	}
 }
 
-func (v *Vlan) getSlaveNIC() (*iface.Link, error) {
-	nics, err := iface.ListLinks(map[string]bool{iface.TypeDevice: true, iface.TypeBond: true})
+func (v *Vlan) getUplink() (*iface.Link, error) {
+	ifaces, err := iface.ListLinks(map[string]bool{iface.TypeDevice: true, iface.TypeBond: true})
 	if err != nil {
-		return nil, fmt.Errorf("get NICs failed, error: %w", err)
+		return nil, fmt.Errorf("get uplink of VLAN %s failed, error: %w", v.name, err)
 	}
 
 	var number int
-	var slaveNIC string
-	for _, n := range nics {
-		if n.LinkAttrs().MasterIndex == v.bridge.Index() {
-			slaveNIC = n.Name()
+	var uplink *iface.Link
+	for _, i := range ifaces {
+		if i.Attrs().MasterIndex == v.bridge.Index {
+			uplink = i
 			number++
 		}
 	}
 	if number > 1 {
-		return nil, fmt.Errorf("the number of slave NICs can not be over one, actual numbers: %d", number)
+		return nil, fmt.Errorf("the number of uplinks of %s can not be over one, actual numbers: %d", v.name, number)
 	}
-
 	if number == 0 {
-		return nil, SlaveNotFoundError{fmt.Errorf("slave of %s not found", v.bridge.Name())}
+		return nil, fmt.Errorf("uplink of %s not found", v.name)
 	}
 
-	return iface.GetLinkByName(slaveNIC)
+	return uplink, nil
 }
 
-func GetVlan(mgmtNetwork network.Network) (*Vlan, error) {
-	v := NewVlan(nil, mgmtNetwork, nil)
+func GetVlan(name string) (*Vlan, error) {
+	v := NewVlan(name, nil)
 	if err := v.bridge.Fetch(); err != nil {
 		return nil, err
 	}
 
-	nic, err := v.getSlaveNIC()
+	uplink, err := v.getUplink()
+	if err != nil {
+		return nil, err
+	}
+	v.uplink = uplink
+
+	v.vids, err = v.uplink.ListBridgeVlan()
 	if err != nil {
 		return nil, err
 	}
-	v.nic = nic
 
 	return v, nil
 }
 
-func (v *Vlan) Setup(nic string) error {
-	klog.Info("start to setup VLAN network")
+func (v *Vlan) Setup(l *iface.Link) error {
 	// ensure bridge and get NIC
 	if err := v.bridge.Ensure(); err != nil {
-		return fmt.Errorf("ensure bridge %s failed, error: %w", v.bridge.Name(), err)
-	}
-	l, err := iface.GetLinkByName(nic)
-	if err != nil {
-		return fmt.Errorf("get NIC %s failed, error: %w", nic, err)
-	}
-	if l.LinkAttrs().OperState == netlink.OperDown {
-		return fmt.Errorf("NIC %s is down", l.Name())
-	}
-
-	// Use the same NIC with the flannel network
-	if l.Index() == v.mgmtNetwork.NIC().Index() {
-		if err := v.bridge.SyncIPv4Addr(l); err != nil {
-			return err
-		}
-		if err := v.bridge.ToLink().AddRoutes(l); err != nil {
-			return err
-		}
+		return fmt.Errorf("ensure bridge %s failed, error: %w", v.bridge.Name, err)
 	}
 
 	// set master
```

Comment:
vlan to:
```
  set uplink's master as bridge
  (loop) set all vids to uplink's vlan filter
```  

link's `SetMaster` should not combine `set master` and `set vlan filter`